### PR TITLE
DA-3097: Implement: NPH Inventory Process

### PR DESCRIPTION
**Pull request by jmtrujillo75**
## Resolves *[DA-3097](https://precisionmedicineinitiative.atlassian.net/browse/DA-3097)*

## Description of changes/additions

1. Added `StoredSample` model
2. Modified the `Participant.biobank_id` field to be `non-nullable` and Created an Index on the field
3. Added Cron Job entry for NPH Biobank Inventory Process 
4. Renamed `tools/tool_libs/study_nph_biobank_file_export.py` to `tools/tool_libs/study_nph_biobank_process.py`


## Tests
Tested the changes on `localhost`

1. Generate Test Participant Data
```zsh
$ python ./data_gen/generators/temp_participants_for_inventory_process.py
```
Generated Participants
```sql
SELECT * from participant p;
id|created            |modified           |ignore_flag|disable_flag|disable_reason                                                                                                                                                                                                                                                 |biobank_id|research_id|
--+-------------------+-------------------+-----------+------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+-----------+
 1|2023-01-31 20:48:36|2023-01-31 20:48:36|          1|           0|                                                                                                                                                                                                                                                               |  54789258|    7105211|
 2|2023-01-31 20:48:36|2023-01-31 20:48:36|          0|           1|UhiMlfyranUwHIWBSnoFDbNibZFzMpWdySkNxDLKxNrjTGbqMXsVcBcqFaOdeDsuMscTmuWErVzzObZteAqVosEVjttIanaUatwCrmdXMEFpXpDchAAOeTgBQdMnGpnasOjnXUZIgiUlZRtZYttJljTCtcARjLqoIgmQlkyTXzYBRoItvIfFHvwjAMTHyCOPXyIaReOMiOmDeWNzAPYJCSjUPhfGcYrUfHPJWTqUAsUGxOdbJDIhLcelBfjBxbH|  54789632|    6438188|
 3|2023-01-31 20:48:36|2023-01-31 20:48:36|          0|           0|                                                                                                                                                                                                                                                               |  54789678|    8828999|
 4|2023-01-31 20:48:36|2023-01-31 20:48:36|          1|           0|                                                                                                                                                                                                                                                               |  54789624|    7654096|
 5|2023-01-31 20:48:36|2023-01-31 20:48:36|          1|           0|                                                                                                                                                                                                                                                               |  54789256|    5906901|
 6|2023-01-31 20:48:36|2023-01-31 20:48:36|          1|           1|hAuuOHsvUWTgKDbCmtErmvOYPPHTXayUxStFElrVxmJOYHQFMtqBwrKptiwsGQLbHCitLsAsVgywgfBVwljoliRoyoyEoHPwigZSifjrUzUQumRIuMiKFKvVdDLOOPaRLNbnvZUuFRNwZhIfcRusqgyUyozCsRNimOUlLPHaeegYgvmkAowRyGZIArCvIYMUCfvsWkwHYJLBoeqjKtUaPrBWygFCkucTFrWuFHQWGJgNGktqUAEcBeodcqdNJKr|  54789785|    3168498|
 7|2023-01-31 20:48:37|2023-01-31 20:48:37|          0|           1|NCCOTNVpDqOsfhPNtTcTWfuDBGUpYaYVczoudHLlGxbTWzLzirvzdLfUhSJKCiYMNvhWPgpWChOggitzRgTSiCeHymfkNhidEOFlpEvPAxgNBHtedthBOKMcWVjwXCGLlWCVJrXkBurnTuzDkWVRMRcwleFiBHZBoOoyTumbDJKSOhaxySnOWfJPOGfEJpiyNEIsgHkoSSGYyETgqyJsmYCxykZjHOnCaytYBqREFupogFsobubxvPHMQjdFsuW|  54789683|    2557096|
 8|2023-01-31 20:48:37|2023-01-31 20:48:37|          0|           1|iKLMAbbMHcmjnxGJpyaoZIbWyJWrJPoeNWYTZQlZAxYUAXLqoHeHnFhrtJLakpinWiGkMIPuJgfHtvcftcfjezZeILaIityHXRLNFocHNjYepbuJPdsVrTVoThfDSDWsshcXyngJvLNgObGKFVEfJiNtKwvLFGtWobmcFAmAQQROFBoGmZoFmCWZPomaHPbQBwvvNomAJFJzhGWGTkQEPhcXYEmCFiDvJIasbjZJXzmIIFcdazHjpjEmQzzzfWj|  54789123|    8459739|
 9|2023-01-31 20:48:37|2023-01-31 20:48:37|          0|           0|                                                                                                                                                                                                                                                               |  54789487|    6530550|
```

2. Store the NPH Inventory file in the local storage
```zsh
(venv) rdr_service $ stat ../tests/.test_storage/stable-nph-sample-data-biobank/nph_inventory_process/sample_nph_biobank_inventory_file.csv 
16777234 7481242 -rw-r--r-- 1 trujiljm staff 0 6534 "Jan 29 17:54:14 2023" "Jan 29 17:54:14 2023" "Jan 29 17:55:12 2023" "Jan 29 17:54:14 2023" 4096 16 0 ../tests/.test_storage/stable-nph-sample-data-biobank/nph_inventory_process/sample_nph_biobank_inventory_file.csv

# Total No. of Rows
(venv) rdr_service $ wc -l ../tests/.test_storage/stable-nph-sample-data-biobank/nph_inventory_process/sample_nph_biobank_inventory_file.csv
      99 ../tests/.test_storage/stable-nph-sample-data-biobank/nph_inventory_process/sample_nph_biobank_inventory_file.csv
```

3. Run the import job locally
```zsh
$ python -m tools study_nph_biobank_process --project all-of-us-rdr-stable study_nph_biobank_inventory_file_import
CPUs: 10.


finished.
```

4. Verify the `nph.stored_sample` table
```sql
SELECT COUNT(*) nph.stored_sample s;
COUNT(*)|
--------+
      99|

SELECT * FROM nph.stored_sample LIMIT 10;
id |created            |modified           |biobank_modified   |biobank_id|sample_id|lims_id    |status|disposition|
---+-------------------+-------------------+-------------------+----------+---------+-----------+------+-----------+
100|2023-01-31 21:07:56|2023-01-31 21:07:56|2021-08-18 18:29:47|  54789632|414711031|22630800012|     0|           |
101|2023-01-31 21:07:56|2023-01-31 21:07:56|2021-08-18 18:29:47|  54789632|414711063|22630800013|     0|           |
102|2023-01-31 21:07:56|2023-01-31 21:07:56|2021-08-18 18:29:47|  54789632|414711095|22630800014|     0|           |
103|2023-01-31 21:07:56|2023-01-31 21:07:56|2021-08-18 18:29:47|  54789632|414711025|22630800015|     0|           |
104|2023-01-31 21:07:56|2023-01-31 21:07:56|2021-08-18 18:29:47|  54789632|414711097|22630800016|     0|           |
105|2023-01-31 21:07:56|2023-01-31 21:07:56|2021-08-18 18:29:47|  54789632|414711057|22630800017|     0|           |
106|2023-01-31 21:07:56|2023-01-31 21:07:56|2021-08-18 18:29:47|  54789632|414711065|22630800018|     0|           |
107|2023-01-31 21:07:56|2023-01-31 21:07:56|2021-08-18 18:29:47|  54789632|414706453|22630800019|     0|           |
108|2023-01-31 21:07:56|2023-01-31 21:07:56|2021-08-18 18:29:47|  54789678|414711016|22630800020|     0|           |
109|2023-01-31 21:07:56|2023-01-31 21:07:56|2021-08-18 18:29:47|  54789678|414733707|22630800021|     2|           |
```

[DA-3097]: https://precisionmedicineinitiative.atlassian.net/browse/DA-3097?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by kenny-skaggs — APPROVED** [review]

File: rdr_service/model/study_nph.py [hunk 1]
```diff
@@ -279,3 +282,21 @@ class SampleExport(NphBase):
...
     sample_update_id = Column(BigInteger, ForeignKey("sample_update.id"))
+
+
+class StoredSample(NphBase):
+    __tablename__ = "stored_sample"
+
+    id = Column("id", BigInteger, autoincrement=True, primary_key=True)
+    created = Column(UTCDateTime)
+    modified = Column(UTCDateTime)
+    biobank_modified = Column(UTCDateTime)
```
Comment (j-kanuch) [inline rdr_service/model/study_nph.py:296]: This needs an ignore_flag field like the other tables.

File: rdr_service/tools/tool_libs/study_nph_biobank_process.py [hunk 2]
```diff
@@ -32,13 +34,32 @@ def run(self):
...
+
 
+class StudyNphBioBankInventoryFileImport(ToolBase):
+
+    def __init__(self, args, gcp_env: GCPEnvConfigObject):
+        super().__init__(args, gcp_env)
+
+    def run(self):
+        if self.args.project == 'all-of-us-rdr-prod':
+            _logger.error(f'Participant generator cannot be used on project: {self.args.project}')
```
Comment (j-kanuch) [inline rdr_service/tools/tool_libs/study_nph_biobank_process.py]: nit: this error message should be updated to match the tool name.

File: rdr_service/model/study_nph_enums.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+from protorpc import messages
+
+
+class StoredSampleStatus(messages.Enum):
+    """Status of Genomic Set"""
```
Comment (j-kanuch) [inline rdr_service/model/study_nph_enums.py]: nit: this docstring should be updated.

File: rdr_service/model/study_nph.py [hunk 4]
```diff
@@ -279,3 +282,21 @@ class SampleExport(NphBase):
...
+    biobank_id = Column(BigInteger, ForeignKey("participant.biobank_id"))
+    sample_id = Column(BigInteger)
```
Comment (j-kanuch) [inline rdr_service/model/study_nph.py]: Can you set `index=True` on the `sample_id`? I have a feeling we'll need it in the long run.

**Review by j-kanuch — APPROVED** [review]

**Pull request merged**